In [1]:
import numpy as np
import pandas as pd

In [32]:
data = pd.read_csv("horse.csv")
data.head()

surgery    age  hospital_number  rectal_temp  pulse  respiratory_rate  \
0      no  adult           530101         38.5   66.0              28.0   
1     yes  adult           534817         39.2   88.0              20.0   
2      no  adult           530334         38.3   40.0              24.0   
3     yes  young          5290409         39.1  164.0              84.0   
4      no  adult           530255         37.3  104.0              35.0   

  temp_of_extremities peripheral_pulse mucous_membrane capillary_refill_time  \
0                cool          reduced             NaN            more_3_sec   
1                 NaN              NaN   pale_cyanotic            less_3_sec   
2              normal           normal       pale_pink            less_3_sec   
3                cold           normal   dark_cyanotic            more_3_sec   
4                 NaN              NaN   dark_cyanotic            more_3_sec   

   ... packed_cell_volume total_protein abdomo_appearance abdomo_protein  \
0  ...               45.0           8.4               NaN            NaN   
1  ...               50.0          85.0            cloudy            2.0   
2  ...               33.0           6.7               NaN            NaN   
3  ...               48.0           7.2     serosanguious            5.3   
4  ...               74.0           7.4               NaN            NaN   

      outcome  surgical_lesion lesion_1 lesion_2  lesion_3  cp_data  
0        died               no    11300        0         0       no  
1  euthanized               no     2208        0         0       no  
2       lived               no        0        0         0      yes  
3        died              yes     2208        0         0      yes  
4        died               no     4300        0         0       no  

[5 rows x 28 columns]

In [11]:
data.shape

(299, 28)

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   surgery                299 non-null    object 
 1   age                    299 non-null    object 
 2   hospital_number        299 non-null    int64  
 3   rectal_temp            239 non-null    float64
 4   pulse                  275 non-null    float64
 5   respiratory_rate       241 non-null    float64
 6   temp_of_extremities    243 non-null    object 
 7   peripheral_pulse       230 non-null    object 
 8   mucous_membrane        252 non-null    object 
 9   capillary_refill_time  267 non-null    object 
 10  pain                   244 non-null    object 
 11  peristalsis            255 non-null    object 
 12  abdominal_distention   243 non-null    object 
 13  nasogastric_tube       195 non-null    object 
 14  nasogastric_reflux     193 non-null    object 
 15  nasoga

In [13]:
data.columns

Index(['surgery', 'age', 'hospital_number', 'rectal_temp', 'pulse',
       'respiratory_rate', 'temp_of_extremities', 'peripheral_pulse',
       'mucous_membrane', 'capillary_refill_time', 'pain', 'peristalsis',
       'abdominal_distention', 'nasogastric_tube', 'nasogastric_reflux',
       'nasogastric_reflux_ph', 'rectal_exam_feces', 'abdomen',
       'packed_cell_volume', 'total_protein', 'abdomo_appearance',
       'abdomo_protein', 'outcome', 'surgical_lesion', 'lesion_1', 'lesion_2',
       'lesion_3', 'cp_data'],
      dtype='object')

In [15]:
data.describe().T

count          mean           std       min       25%  \
hospital_number        299.0  1.087733e+06  1.532032e+06  518476.0  528904.0   
rectal_temp            239.0  3.816862e+01  7.337444e-01      35.4      37.8   
pulse                  275.0  7.200000e+01  2.864622e+01      30.0      48.0   
respiratory_rate       241.0  3.046058e+01  1.766610e+01       8.0      18.0   
nasogastric_reflux_ph   53.0  4.707547e+00  1.982311e+00       1.0       3.0   
packed_cell_volume     270.0  4.630741e+01  1.043674e+01      23.0      38.0   
total_protein          266.0  2.427444e+01  2.736419e+01       3.3       6.5   
abdomo_protein         101.0  3.039604e+00  1.967947e+00       0.1       2.0   
lesion_1               299.0  3.659709e+03  5.408472e+03       0.0    2111.5   
lesion_2               299.0  9.052843e+01  6.506371e+02       0.0       0.0   
lesion_3               299.0  7.387960e+00  1.277498e+02       0.0       0.0   

                            50%        75%        max  
hospital_number        530301.0  534736.00  5305629.0  
rectal_temp                38.2      38.50       40.8  
pulse                      64.0      88.00      184.0  
respiratory_rate           25.0      36.00       96.0  
nasogastric_reflux_ph       5.0       6.50        7.5  
packed_cell_volume         45.0      52.00       75.0  
total_protein               7.5      56.75       89.0  
abdomo_protein              2.3       3.90       10.1  
lesion_1                 2322.0    3209.00    41110.0  
lesion_2                    0.0       0.00     7111.0  
lesion_3                    0.0       0.00     2209.0

In [16]:
data.isnull().sum()

surgery                    0
age                        0
hospital_number            0
rectal_temp               60
pulse                     24
respiratory_rate          58
temp_of_extremities       56
peripheral_pulse          69
mucous_membrane           47
capillary_refill_time     32
pain                      55
peristalsis               44
abdominal_distention      56
nasogastric_tube         104
nasogastric_reflux       106
nasogastric_reflux_ph    246
rectal_exam_feces        102
abdomen                  118
packed_cell_volume        29
total_protein             33
abdomo_appearance        165
abdomo_protein           198
outcome                    0
surgical_lesion            0
lesion_1                   0
lesion_2                   0
lesion_3                   0
cp_data                    0
dtype: int64

as ween above we get that there are number of the missing value inside the data.

In [28]:
data = pd.read_csv("horse.csv",header=None,na_values="?")#to get the column name in the index wise like 1,2,3,34 etc.

In [19]:
# summarize the number of rows with missing values for each column
for i in range(data.shape[1]):
    # count number of rows with missing values
    n_miss = data[[i]].isnull().sum()
    perc = n_miss / data.shape[0] * 100
    print('> %d, Missing: %d (%.1f%%)' % (i, n_miss, perc))

> 0, Missing: 0 (0.0%)
> 1, Missing: 0 (0.0%)
> 2, Missing: 0 (0.0%)
> 3, Missing: 60 (20.0%)
> 4, Missing: 24 (8.0%)
> 5, Missing: 58 (19.3%)
> 6, Missing: 56 (18.7%)
> 7, Missing: 69 (23.0%)
> 8, Missing: 47 (15.7%)
> 9, Missing: 32 (10.7%)
> 10, Missing: 55 (18.3%)
> 11, Missing: 44 (14.7%)
> 12, Missing: 56 (18.7%)
> 13, Missing: 104 (34.7%)
> 14, Missing: 106 (35.3%)
> 15, Missing: 246 (82.0%)
> 16, Missing: 102 (34.0%)
> 17, Missing: 118 (39.3%)
> 18, Missing: 29 (9.7%)
> 19, Missing: 33 (11.0%)
> 20, Missing: 165 (55.0%)
> 21, Missing: 198 (66.0%)
> 22, Missing: 0 (0.0%)
> 23, Missing: 0 (0.0%)
> 24, Missing: 0 (0.0%)
> 25, Missing: 0 (0.0%)
> 26, Missing: 0 (0.0%)
> 27, Missing: 0 (0.0%)


here we seen that how many percentage the missing value is present inside the column.

# imputing the missing value:

- we can do this by using the mean 
- we can also do this with median.
- we can also do this with the mode.

In [29]:
from numpy import isnan
from pandas import read_csv
from sklearn.impute import SimpleImputer

In [36]:
data = pd.read_csv("horse.csv")

In [33]:
data = data.dropna(axis=0)

In [34]:
data.head()

surgery    age  hospital_number  rectal_temp  pulse  respiratory_rate  \
10      yes  adult           528548         38.1   66.0              12.0   
54       no  adult           529461         40.3  114.0              36.0   
64      yes  adult           529667         39.0   64.0              36.0   
91       no  adult           529461         40.3  114.0              36.0   
104     yes  adult           527563         37.8   52.0              24.0   

    temp_of_extremities peripheral_pulse mucous_membrane  \
10                 cool          reduced      bright_red   
54                 cool          reduced     normal_pink   
64                 cool           normal   pale_cyanotic   
91                 cool          reduced     normal_pink   
104              normal          reduced       pale_pink   

    capillary_refill_time  ... packed_cell_volume total_protein  \
10             less_3_sec  ...               44.0           6.0   
54             more_3_sec  ...               57.0           8.1   
64             more_3_sec  ...               44.0           7.5   
91             more_3_sec  ...               57.0           8.1   
104            less_3_sec  ...               48.0           6.6   

    abdomo_appearance abdomo_protein     outcome  surgical_lesion lesion_1  \
10             cloudy            3.6       lived              yes     2124   
54      serosanguious            4.5  euthanized              yes     7400   
64      serosanguious            5.0       lived              yes     2113   
91      serosanguious            4.5        died              yes     3205   
104             clear            3.7        died              yes     5400   

    lesion_2  lesion_3  cp_data  
10         0         0      yes  
54         0         0      yes  
64         0         0      yes  
91         0         0      yes  
104        0         0       no  

[5 rows x 28 columns]

In [35]:
data.shape

(6, 28)

    Initially we have the data shape 299*28 and now we seen above we get that data shape is 6*28 beacuse we drop the rows which is having the null values.

In [48]:
data = pd.read_csv("horse.csv")

In [49]:
data.head(3)

surgery    age  hospital_number  rectal_temp  pulse  respiratory_rate  \
0      no  adult           530101         38.5   66.0              28.0   
1     yes  adult           534817         39.2   88.0              20.0   
2      no  adult           530334         38.3   40.0              24.0   

  temp_of_extremities peripheral_pulse mucous_membrane capillary_refill_time  \
0                cool          reduced             NaN            more_3_sec   
1                 NaN              NaN   pale_cyanotic            less_3_sec   
2              normal           normal       pale_pink            less_3_sec   

   ... packed_cell_volume total_protein abdomo_appearance abdomo_protein  \
0  ...               45.0           8.4               NaN            NaN   
1  ...               50.0          85.0            cloudy            2.0   
2  ...               33.0           6.7               NaN            NaN   

      outcome  surgical_lesion lesion_1 lesion_2  lesion_3  cp_data  
0        died               no    11300        0         0       no  
1  euthanized               no     2208        0         0       no  
2       lived               no        0        0         0      yes  

[3 rows x 28 columns]

In [42]:
data.isnull().mean()

surgery                  0.000000
age                      0.000000
hospital_number          0.000000
rectal_temp              0.200669
pulse                    0.080268
respiratory_rate         0.193980
temp_of_extremities      0.187291
peripheral_pulse         0.230769
mucous_membrane          0.157191
capillary_refill_time    0.107023
pain                     0.183946
peristalsis              0.147157
abdominal_distention     0.187291
nasogastric_tube         0.347826
nasogastric_reflux       0.354515
nasogastric_reflux_ph    0.822742
rectal_exam_feces        0.341137
abdomen                  0.394649
packed_cell_volume       0.096990
total_protein            0.110368
abdomo_appearance        0.551839
abdomo_protein           0.662207
outcome                  0.000000
surgical_lesion          0.000000
lesion_1                 0.000000
lesion_2                 0.000000
lesion_3                 0.000000
cp_data                  0.000000
dtype: float64

In [53]:
data.fillna(data.mean())

surgery    age  hospital_number  rectal_temp  pulse  respiratory_rate  \
0        no  adult           530101    38.500000   66.0              28.0   
1       yes  adult           534817    39.200000   88.0              20.0   
2        no  adult           530334    38.300000   40.0              24.0   
3       yes  young          5290409    39.100000  164.0              84.0   
4        no  adult           530255    37.300000  104.0              35.0   
..      ...    ...              ...          ...    ...               ...   
294     yes  adult           533886    38.168619  120.0              70.0   
295      no  adult           527702    37.200000   72.0              24.0   
296     yes  adult           529386    37.500000   72.0              30.0   
297     yes  adult           530612    36.500000  100.0              24.0   
298     yes  adult           534618    37.200000   40.0              20.0   

    temp_of_extremities peripheral_pulse mucous_membrane  \
0                  cool          reduced             NaN   
1                   NaN              NaN   pale_cyanotic   
2                normal           normal       pale_pink   
3                  cold           normal   dark_cyanotic   
4                   NaN              NaN   dark_cyanotic   
..                  ...              ...             ...   
294                cold              NaN   pale_cyanotic   
295                cool        increased   pale_cyanotic   
296                cold          reduced   pale_cyanotic   
297                cool          reduced       pale_pink   
298                 NaN              NaN             NaN   

    capillary_refill_time  ... packed_cell_volume total_protein  \
0              more_3_sec  ...               45.0      8.400000   
1              less_3_sec  ...               50.0     85.000000   
2              less_3_sec  ...               33.0      6.700000   
3              more_3_sec  ...               48.0      7.200000   
4              more_3_sec  ...               74.0      7.400000   
..                    ...  ...                ...           ...   
294            more_3_sec  ...               55.0     65.000000   
295            more_3_sec  ...               44.0     24.274436   
296            less_3_sec  ...               60.0      6.800000   
297            less_3_sec  ...               50.0      6.000000   
298                   NaN  ...               36.0     62.000000   

    abdomo_appearance abdomo_protein     outcome  surgical_lesion lesion_1  \
0                 NaN       3.039604        died               no    11300   
1              cloudy       2.000000  euthanized               no     2208   
2                 NaN       3.039604       lived               no        0   
3       serosanguious       5.300000        died              yes     2208   
4                 NaN       3.039604        died               no     4300   
..                ...            ...         ...              ...      ...   
294               NaN       3.039604  euthanized               no     3205   
295     serosanguious       3.300000  euthanized              yes     2208   
296               NaN       3.039604        died              yes     3205   
297     serosanguious       3.400000       lived              yes     2208   
298             clear       1.000000  euthanized               no     6112   

    lesion_2  lesion_3  cp_data  
0          0         0       no  
1          0         0       no  
2          0         0      yes  
3          0         0      yes  
4          0         0       no  
..       ...       ...      ...  
294        0         0       no  
295        0         0      yes  
296        0         0       no  
297        0         0      yes  
298        0         0       no  

[299 rows x 28 columns]

In [51]:
data.head(3)

surgery    age  hospital_number  rectal_temp  pulse  respiratory_rate  \
0      no  adult           530101         38.5   66.0              28.0   
1     yes  adult           534817         39.2   88.0              20.0   
2      no  adult           530334         38.3   40.0              24.0   

  temp_of_extremities peripheral_pulse mucous_membrane capillary_refill_time  \
0                cool          reduced             NaN            more_3_sec   
1                 NaN              NaN   pale_cyanotic            less_3_sec   
2              normal           normal       pale_pink            less_3_sec   

   ... packed_cell_volume total_protein abdomo_appearance abdomo_protein  \
0  ...               45.0           8.4               NaN            NaN   
1  ...               50.0          85.0            cloudy            2.0   
2  ...               33.0           6.7               NaN            NaN   

      outcome  surgical_lesion lesion_1 lesion_2  lesion_3  cp_data  
0        died               no    11300        0         0       no  
1  euthanized               no     2208        0         0       no  
2       lived               no        0        0         0      yes  

[3 rows x 28 columns]

In [52]:
data.shape

(299, 28)

In [ ]:

# replacing missing values in quantity
# column with mean of that column
data['quantity'] = data['quantity'].fillna(data['quantity'].mean())
  
# replacing missing values in price column
# with median of that column
data['price'] = data['price'].fillna(data['price'].median())
  
# replacing missing values in bought column with
# standard deviation of that column
data['bought'] = data['bought'].fillna(data['bought'].std())
  
# replacing missing values in forenoon  column with
# minimum number of that column
data['forenoon'] = data['forenoon'].fillna(data['forenoon'].min())
  
# replacing missing values in afternoon  column with 
# maximum number of that column
data['afternoon'] = data['afternoon'].fillna(data['afternoon'].max())
  
print(Data)